ceno_zkvm/src/scheme/verifier.rs

```diff
@@ -140,43 +140,63 @@ impl<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>> ZKVMVerifier<E, PCS>
         let point_eval = PointAndEval::default();
         let mut transcripts = transcript.fork(self.vk.circuit_vks.len());
 
-        for (name, (i, opcode_proof)) in vm_proof.opcode_proofs {
-            let transcript = &mut transcripts[i];
+        // For each opcode, include the num_instances
+        // into its corresponding fork of the transcript.
+        for ((name, _), transcript) in self
+            .vk
+            .circuit_vks
+            .iter() // Sorted by key.
+            .zip_eq(transcripts.iter_mut())
+        {
+            // get num_instances from opcode proof
+            let opcode_result = vm_proof.opcode_proofs.get(name);
+            let num_instances = opcode_result.map(|(_, p)| p.num_instances).unwrap_or(0);
+            if opcode_result.is_some() {
+                transcript.append_field_element(&E::BaseField::from(num_instances as u64));
```

Comment:
It would protect against more potential attacks to do this **before** the `alpha, beta read_challenge` above. It’s before the transcript fork, so just a sequence of append to the root transcript.